python/pyspark/sql/tests/test_arrow.py

```diff
@@ -191,6 +191,32 @@ def test_pandas_round_trip(self):
         pdf_arrow = df.toPandas()
         assert_frame_equal(pdf_arrow, pdf)
 
+    def test_pandas_self_destruct(self):
+        import pyarrow as pa
+        rows = 2 ** 16
+        cols = 8
+        df = self.spark.range(0, rows).select(*[rand() for _ in range(cols)])
+        expected_bytes = rows * cols * 8
+        with self.sql_conf({"spark.sql.execution.arrow.pyspark.selfDestruct.enabled": True}):
+            # We hold on to the table reference here, so if self destruct didn't work, then
+            # there would be 2 copies of the data (one in Arrow, one in Pandas), both
+            # tracked by the Arrow allocator
```

Comment:
Oh, I missed this comment that you need the Table reference. Otherwise it could be freed and then we lose track of memory allocated right? If that's the case, what about just calling `_collect_as_arrow(_force_split_batches=True)` here and perform batches -> table -> pandas directly?

Reply:
Yes - we need the table reference. If self destruct worked, then it holds effectively 0 memory; if it didn't work, then it holds a copy of the dataframe. 

If we call to_pandas inside the test itself, wouldn't we then be testing Arrow directly, instead of testing how Arrow is used inside Spark?

Reply:
Not really, it's not ideal but it would still test that splitting and reallocating the columns in `_collect_as_arrow()` will allow for self destruction, which is most of additions here. The rest is just calling `to_pandas()` with the right options. I would recommend testing something like this:

```python
with self.sql_conf({"spark.sql.execution.arrow.pyspark.selfDestruct.enabled": True}):
    # get initial pa.total_allocated_bytes()
    batches = _collect_as_arrow(_force_split_batches=True)
    table = # convert batches to table as done in toPandas()
    pdf = table.to_pandas(self_destruct=True, split_blocks=True, use_threads=False)
    # get after pa.total_allocated_bytes() and check difference is expected

    # Call with the full code path and compare resulting DataFrame for good measure
    assert_frames_equal(df.toPandas(), pdf)
```

That should be sufficient testing, wdyt?

Reply:
Makes sense to me - I'll try to update this soon. Thanks for being patient with me here.

Also looks like the 3.1 branch was cut already, so I need to update the flags here to reference 3.2.